demo/mpfao/mpfao.c

```diff
@@ -222,10 +228,10 @@ int main(int argc, char **argv) {
   ierr = TDyDistributeDM(&dm); CHKERRQ(ierr);
 
   // Setup problem parameters
-  TDy  tdy;
   PetscReal gravity[3];
 
-  ierr = TDyCreateWithDM(dm,&tdy); CHKERRQ(ierr);
+  ierr = TDySetDM(dm,tdy); CHKERRQ(ierr);
+  ierr = TDyAllocate(tdy); CHKERRQ(ierr);
```

Comment:
For some reason I'm having trouble finding this function. What does it do? Could it be merged into `TDySetup` below, or are you doing something in between here and that call that requires it to be here?

Reply:
Ahh, I'm guessing from your description that `TDyAllocate` allocates space for parameter arrays that allow you to select functions for boundary conditions, source terms, etc. If setting the DM is the last precondition for allocating these parameter arrays, and we disallow the DM to be set for a TDy more than once, perhaps we can do this allocation at the end of `TDySetDM`.

In general, I'm trying to push for a minimum of expert knowledge in properly setting up a TDy instance. The more of these types of calls we eliminate, the less we have to explain to a basic user what's going on (and the less chance we have of messing things up).

Of course, there's a floor to the simplicity we can bestow on this process. But I'm always interested in where that floor actually is. :-)

Reply:
I agree.  I have renamed TDySetDM -> TDySetupDiscretization and moved the portion of TDyAllocate that calculates the geometry within.  I then embed the call to TDyAllocate at the end of TDySetupDiscretization.  The calls to TDyAllocate have been removed from the demos.